### PR TITLE
Chore: Add type info for grafana boot data

### PR DESCRIPTION
**Pull request by kaydelaney**
**What this PR does / why we need it**:
Adds type info to reduce usage of `any` with boot data and related types.

### CODE REVIEW

File: packages/grafana-data/src/types/config.ts [hunk 1]
```diff
@@ -85,6 +86,45 @@ export type OAuth =
  */
 export type OAuthSettings = Partial<Record<OAuth, { name: string }>>;
 
+/** Current user info included in bootData
+ *
+ * @public
```
Comment (ryantxu) [inline packages/grafana-data/src/types/config.ts:91]: @mckn @torkelo -- what are your thoughs on this as a "Public" API -- I guess it is accidentally

The backend API is periodically adding/changing things with this kind of object -- that I don't really want to be commitd to as a public API

Any thoughts on how to best communicate that you can count on:
* login
* email
* name
But that
* helpFlags1
* hasEditPermissionsInFolder 
* etc
are not to be counted on for long term

File: public/app/types/accessControl.ts [hunk 2]
```diff
@@ -1,12 +1,10 @@
 /**
  * UserPermission is a map storing permissions in a form of
  * {
- *   action: { scope: scope }
+ *   action: enabled;
```
Comment (leventebalogh) [inline public/app/types/accessControl.ts:4]: **Suggestion:** should we use `true` instead as it is a boolean?

```suggestion
 *   action: true;
```

File: public/app/types/accessControl.ts [hunk 3]
```diff
@@ -1,12 +1,10 @@
...
  * }
  */
-export type UserPermission = {
-  [key: string]: { [key: string]: string };
-};
+export type UserPermission = Record<string, boolean>;
```
Comment (leventebalogh) [inline public/app/types/accessControl.ts:7]: **Question:** just out of curiosity, why has the structure of this change?

File: packages/grafana-data/src/types/config.ts (hunk 1 above)
Comment (leventebalogh) [reply packages/grafana-data/src/types/config.ts:91]: Do we really want to expose all these information to the plugins?

I would probably say that we would like to provide more specific APIs to access these (or a subset of them) from plugins. Of course it's a bit hard to make it truly internal until it lives in `@grafana/data`.

On the other hand as it is indirectly exposed through the `GrafanaConfig`, does `@public` make any difference here?

Comment (kaydelaney) [reply packages/grafana-data/src/types/config.ts:91]: Yeah that's fair, maybe we could expose a subset of the interface, and then the "full" interface extending it is kept private? Or we could use jsdoc annotations to specify these properties shouldn't be used or depended on?

Comment (mckn) [reply packages/grafana-data/src/types/config.ts:91]: Yeah, it is part of the `BootData` and hence it is part of that API and should have the same release tag as that.

Maybe we can do something in lines with what @kaydelaney suggests?

File: packages/grafana-data/src/types/config.ts [hunk 4]
```diff
@@ -85,6 +86,43 @@ export type OAuth =
  */
 export type OAuthSettings = Partial<Record<OAuth, { name: string; icon?: string }>>;
 
+/** Current user info included in bootData
+ *
+ * @public
+ */
+export interface CurrentUserDTO {
```
Comment (kaydelaney) [inline packages/grafana-data/src/types/config.ts:93]: Any thoughts on which of these should/shouldn't be public? I've moved `helpFlags1` and `hasEditPermissionsInFolder` to a separate interface that extends this one, but perhaps there are others? @torkelo @mckn

**Review by torkelo — COMMENTED** [review]
Hm.. not sure we should encurage usage of config.bootData , it's a hack/smell

not sure this should be ever used directly, boot data is only to setup the config object. 

Boot data was only intended to bootstrap the config object and contextSrv , so using it directly after feels a bit hacky. 

So accessing config.bootData.user feels wrong, either we should add user to config object or make something like contextSrv accessible to plugins. Maybe userSrv? it looks like all uses of config.bootData is to access current user (Which should be done using contextSrv at least historically). But I am open to suggestions for how to expose current user. What is best? 

Adding typing to bootData is useful but not sure how we limit it's use / exposure externally

File: packages/grafana-data/src/types/config.ts (hunk 1 above)
Comment (torkelo) [reply packages/grafana-data/src/types/config.ts:91]: Hm.. not sure we should encurage usage of config.bootData , it's a hack/smell

not sure this should be ever used directly, boot data is only to setup the config object.

Boot data was only intended to bootstrap the config object and contextSrv , so using it directly after feels a bit hacky.

So accessing config.bootData.user feels wrong, either we should add user to config object or make something like contextSrv accessible to plugins. Maybe userSrv? it looks like all uses of config.bootData is to access current user (Which should be done using contextSrv at least historically). But I am open to suggestions for how to expose current user. What is best?

Adding typing to bootData is useful but not sure how we limit it's use / exposure externally

Comment (joshhunt) [reply packages/grafana-data/src/types/config.ts:91]: This is not _currently_ a public API is it? 

Can we just keep this as private/internal?

The data itself has always been just "publicly available" on window, right, so there's no change in actual exposure.

**Review by joshhunt — COMMENTED** [review]
Excellent stuff. If we're focusing on it, would like just a few of the `as BootData` assertions cleaned up while we're at it if they're not necessary.

Other than that, what's the blocker to merging this in? Would making the scope for the new type private resolve our concerns?

**Comment by kaydelaney** [comment]
Changed the types to `@internal`. Okay to merge now? @leventebalogh @torkelo

**Review by ashharrison90 — APPROVED** [review]
lgtm 🙌

**Pull request merged**